### DISCUSSION TITLE
conditional fields to display different content

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#14] [Ideas]
Having a feature that will  display different content based on a condition.

Exemples: 
1) if cover thumbail available, then display thumbail , else show the plot text instead of the thumbnail 
2) some videos have integrated chapters (info label Player.Chapter, Player.ChapterCount, Player.ChapterName) if the video have no chapter kodi will sent back "00" for Player.Chapter, Player.ChapterCount, so if I display these info label it wil show "00"
This feature will allow to do:
If Player.ChapterCount>1, then display something (eg, Chapter xx/yy, {chapter name})  else display nothing

**Comment by mattblovell** [comment]
>     2\. If Player.ChapterCount>1, then display something (eg, Chapter xx/yy, {chapter name})  else display nothing

This is somewhat similar to the handling of "Unknown" for the `artist` special treatment that exists in `audio_text_fields()`.  At present, there is  a hard-coded comparison that occurs against the string "Unknown" and a test checking of the boolean `drop_unknown` from the TOML file.

That approach could be extended to have an `exclude_value` string (or possible a list of strings) to check for.  If the named InfoLabel matches that string (or, if a list proves possible, any entry in the list), then display of that textfield (and the associated label, if present) gets skipped (just as if the InfoLabel value was the empty string).  (Kodi appears inconsistent here, although perhaps the approach is always to provide a value for fields that are intended to be numeric.)

Regarding Point 1, an "alternate" is possible.  Although, for cover art specifically, one would have to either pass info to or get info back from `get_artwork()` to interact properly with the selection of default images.

However, *all* of the above may not be relevant if special treatment via a lookup dictionary (Idea #16) is implemented.  Since the user-specified callback functions would all be in Python, they could use any desired conditions and have any number of fallback approaches that the user might want to implement.  Both of the above (as well as other proposed Ideas) could be implemented with that approach.

**Comment by mattblovell** [comment]
For the second item in the original post, I think (with kodi_panel v1.27) it should now be possible to specify 

    exclude = "00"

for Player.Chapter.  More sophisticated conditional behavior would best be handled by writing a new string manipulation function and adding it to the `STRING_CB` lookup table.

**Comment by mattblovell** [comment]
> if cover thumbail available, then display thumbail , else show the plot text instead of the thumbnail

The above should now be possible, albeit with some Python programming.  One could add a new entry to the `ELEMENT_CB` table and have a `fields` entry refer to it, and completely omit the `thumb` at the top-level of a layout.

The new element-drawing callback function could invoke `get_artwork()`, saving the result and checking the status of the newly-added `_image_default` flag.  That flag gets set if `get_artwork()` had to make use of one of the default files for artwork.  

- If `_image_default` is True, then one could return the plot (as a string for rendering, making use of additional `wrap`, etc keys).  
- If `_image_default` is False, one could then place the artwork as usual.

The above is admittedly non-trivial!  Since I haven't tried implementing it, there could also be a few holes in my description.

For those reasons, I'll try to ponder how easy it would be to add `alt` functionality to the `thumb` entry.

**Comment by nico1080** [comment]
All these possibilities should work for my exemples but is is not exactly the feature I had in mind.

I will try to re-formulate it by explaining what I will need to create in order to have the wanted behaviour:

Create a new callback functions family: `layout_heuristic_CB ` 
 exemple:
```
def chapter_test:
If Player.ChapterCount>1, return "A", else return "B"
```

This `layout_heuristic_CB`  could then be invoked for any item in the layout with a `heuristic `key
```
[[A_LAYOUT.A_DEFAULT.fields]]
name = "MusicPlayer.TrackNumber"
posx = 420
posy = 132
font = "font7S"
fill = "color_7S"
heuristic ={chapter_test}="A"
```
in my exemple  track number would be displayed only if chapter count returned by kodi infolabel is >1, otherwise it would be not displayed

and an other layout:
```
[[A_LAYOUT.A_DEFAULT.fields]]
name = "MusicPlayer.Time"
posx = 420
posy = 132
font = "font7S"
fill = "color_7S"
heuristic ={chapter_test}="B"
```
in my exemple time  would be displayed only if chapter count returned by kodi infolabel is =<1, otherwise it would be not displayed

The combination of these two `[[A_LAYOUT.A_DEFAULT.fields]]` will display either track number or time based on result

One could also imagine more than 2 retrun value for `layout_heuristic_CB ` function that would conditionaly  displaymore layout items. (if "A" display layout1, if "B" display layout2, if "C" display layout3 etc
Or it could be just "dumbed" to true (or true/false) result.

This feature would allow simpler writing of the `setup.toml` file

**Comment by mattblovell** [comment]
> This feature would allow simpler writing of the `setup.toml` file

Ah, I think I understand now.

I've opened Issue #36.  Per what I post there, I don't believe TOML supports exactly the syntax you propose.  

My thought would be to permit a `fields` entry to provide the new keys
- `display_test` and
- `display_if`

where `display_test` is the name of a `STRING_CB` callback function and `display_if` specifies the string match that must occur for the entry to be rendered.

**Reply by nico1080** [reply]
The drawback of this syntax is that `setup.toml` file would need two additionnal keys for one feature, It is not that user friendly as user can ommit a key. but it should stay managable

Upon further though, would it not be better to:

-  make the `display_test `compatible with infoLabels (not limited to callback function)
-  add a `display_ifnot `keys that just reverse the `display_if `

In my exemple for displaying various label based on `Player.chapter` content (different than "00")

It will make the following field

```
[[V_LAYOUT.V_DEFAULT.fields]]
name  = "my custom name"
format_str = "chapter {Player.chapter}/{Player.chaptercount} {Player.chaptername} "
posx  = 340
posy  = 180
font  = "font_tiny"
fill  = "color_7S"
dynamic = 1
display_test= "Player.chapter"
display_ifnot = "00"
```

With this one would not need a custom callback function for simple test --> A lot user friendly.
I think most user will not use custom callback function

**Comment by mattblovell** [comment]
> With this one would not need a custom callback function for simple test

See update posted to Issue #36

**Comment by mattblovell** [comment]
Implemented in kodi_panel v1.29.

**Comment by nico1080** [comment]
I used this function to display timeshift offest if higher than 10s https://github.com/nico1080/kodi_panel/commit/ab5874e9fc6d8f8b8c884cb76107b56676dcdab4

Also for chapter test : https://github.com/nico1080/kodi_panel/commit/60503a5faaba103dca5a401d68c5a4e7afe488b3

**Comment by nico1080** [comment]
another exemple for showing thumb or plot if thumb is empty b56f56c32dc3193b7afe6afa8e20c9b3ebaade13
